### ISSUE TITLE
Release V3?

### ISSUE THREAD

**Issue by udoprog** [crosire/scripthookvdotnet#869]
Hey,

Me and a couple of downstream projects are trying to make use of v3, but since there isn't an official release yet, most projects are still pinned one way or another to v2 and their branches tend to bit rot.

Is there any chance a release might be coming soon? I haven't found any blockers that are still open apart from potentially using a separate `scripts` folder to indicate that V2 and V3 scripts are incompatible.

**Comment by crosire** [comment]
The whole v3 undertaking is in a bit of a dilemma. It spawned out of the hope of cleaning up the scripting API, but after a while it became clear that it is a very bad idea to just break API compatibility, since the majority of scripts written won't be updated by their authors anymore. That's when the loader was modified to be able to load multiple versions of SHVDN simultanously. But it is certainly not a great solution, since people now have to figure out which versions of SHVDN the scripts they want to use use and install them. It is the very same disaster that Python has to fight with (funny coinsidence being that they even have the same version numbers with this problem, Python2 and Python3).
So releasing v3 would mean the package also has to contain v2 for backwards compatibility. Unfortunately the build system cannot build two branches and package them together automatically currently, so this would have to be done manually every time, which isn't great either ...

**Comment by udoprog** [comment]
Is this tracked somewhere?

If not, I'd like to kick off discussion here.

So how about:

* Version the loader between each major version (`ScriptHookVDotNet.asi` and `ScriptHookVDotNet3.asi`) and make sure they can both work at the same time.
* Release v3 under a new namespace (`GTA` to e.g. `GTA3` or `ScriptHookV3`).
* Make v3 load scripts from a different folder (e.g. `./scripts3`).

It would allow major versions of ScriptHookV.NET to bump the .NET framework version if you want to.

Changing namespace makes sure there's a clean break between the two major ABIs. No partially compiling projects. They actually have to be rewritten to target a new ABI, even if that's just a string replace. Downstream projects can mix and match between targeting two different major ABIs at the same time if they want to without having to have two separate workspaces.

The signatures of the entry points would be different across major versions since they come from different namespaces (`GTA.Script` vs `GTA3.Script`). So in theory, they could both exist in the same project if someone wants them to assuming you haven't bumped the .NET version.

**Comment by crosire** [comment]
I took another shot and this and believe I finally managed to come up with a good solution.

I separated the core SHVDN logic from the scripting API and modified the script domain implementation to be able to load and apply multiple API versions to different scripts. This way everything stays as it is, but just works™. So the core logic resides in `ScriptHookVDotNet.asi` and is responsible for managing the .NET context and all the scripts. Then there are two API DLLs `ScriptHookVDotNet2.dll` and `ScriptHookVDotNet3.dll`, which contain only the API that developers can use (one can choose which version to reference in a script project). On loading a script the core logic will check which API version it was built against and bind the correct version to it.

Really nice plus to this solution: The console that was implemented for v3 is now independent from the version, so it can control both v2 and v3 scripts. It also continues to work to simply put all scripts in the "scripts" subdirectory to the GTA folder, no matter which version they were written against.

Code is currently in a separate branch: https://github.com/crosire/scripthookvdotnet/tree/separate_api.
But I will likely merge this into master soon and get rid of the two seperate version branches (`dev_v2` and `dev_v3`) in favor of having it all in one again. @kagikn FYI

**Comment by justalemon** [comment]
>The console that was implemented for v3 is now independent from the version

Nice! I feel that this would benefit people like me that have been using the cheat input field for executing quick commands.

>one can choose which version to reference in a script project

How this is going to work? By checking the dependency of the assembly?

>But I will likely merge this into master soon

If you need some beta testers prior to merging, let me know.

**Comment by crosire** [comment]
> How this is going to work? By checking the dependency of the assembly?
Yeah. The script assemblies contain a reference with the exact version they were build against. So SHVDN now uses that to determine which API assembly to bind to it (by comparing the major version and checking whether the minor version is greater or equal than the one that was used in the script assembly).
This seems to work well, except for source code scripts of course. They will just be compiled with the newest API (so v3). Considering that there are close to no released scripts in source code form this should be fine though.

I'd be happy for people to try this first (a build is here: https://ci.appveyor.com/project/crosire/scripthookvdotnet/build/artifacts). Checked with a couple of v2 and v3 scripts to see that it works in general so far, but haven't done any stress testing yet. So if you run into issues please let me know. Console opens with `F3` (probably need a better key choice for that).

**Comment by udoprog** [comment]
@crosire 

Nice that you're making progress!

I'd like to propose that we make sure that upgrading SHVDN won't break existing solutions _in place_. This happens if we change default behaviors. So there's two cases we probably want to to make sure we cover before moving forward:

**Drop-in scripts targets V2, unless they specify (somehow) that they target V3**
Easiest would still IMO be to have a separate directory. But really, anything works. You can have a dummy pragma at the top of the script that if absent loads it against V2 (e.g. `/* SHVDN_VERSION=3 */`). What we need to consider here is that the users of the scripts seldomly are the ones who _write_ them. And the authors are the ones who need a stable way to target an API version.

**Transitive dependencies to SHVDN**
I don't have an example of this, but it wouldn't be completely out there. The domain solver probably needs to look through the assembly dependencies recursively, and log conflicts (e.g. which assembly is conflicting). We don't want to be able to end up in situations where two major versions of SHVDN can be pulled in ever due to conflicts in their namespaces. Or we do, but then we probably want to bump the namespace (e.g. `GTA3`) and add support for depending on both in the solver.

I hope this was useful! In case I'm misunderstanding something, feel free to correct me.

**Comment by crosire** [comment]
Names can be conflicting (they already are with v2 and v3), since each type is always associated with a specific assembly. So this is not a problem. Assemblies have specified which assembly a specific type is supposed to be from, so it can be resolved correctly and there can be multiple assemblies with the same types in the appdomain. The build above shows that this is indeed working.

Drop-in source code scripts need a solution, yes. Easiest option would be to just force those to use v2 API. Then existing ones will continue to work and can just add a notice to release notes saying that source code scripts are deprecated and one should pre-compile instead.

**Comment by udoprog** [comment]
> Names can be conflicting (they already are with v2 and v3), since each type is always associated with a specific assembly. So this is not a problem. Assemblies have specified which assembly a specific type is supposed to be from, so it can be resolved correctly and there can be multiple assemblies with the same types in the appdomain. The build above shows that this is indeed working.

When I skimmed over the code I only saw one level of indirection, so I just want to be clear that the following works:
* `A` depends on `SHVDN 2.x.x`.
* `B` depends on `SHVDN 3.x.x`.
* `C` depends on `A` and `B`.

Without isolated packages this would be called "dependency hell" in e.g. Java. But I'm not super into .NET dependency resolution, or if and how transitive dependencies are isolated. So what happens in the above scenario? I believe it's possible to alias dependencies, but can you do it transitively from `C` to avoid linking breakage?

With that lack of knowledge, what I propose is that we make the above situation either error, by checking C's manifest recursively. Or support it through some kind of namespacing.

**Comment by crosire** [comment]
This (type resolving) is all happening at compile time, so nothing SHVDN has to worry about. You can have a C depend on A and B as long as you don't mix direct access to the different SHVDN versions in the same source file (works fine if you go through an indirection via A and B). That would simply fail to compile for obvious reasons.

**Comment by udoprog** [comment]
Ah, I see. Thanks!

**Comment by kagikn** [comment]
@crosire If we decide to port v3 to .NET core, how we manage SHVDN scripts? .NET Core has performance advantages such as `Span` and SIMD and [continues to develop unlike .NET Framework](https://devblogs.microsoft.com/dotnet/net-core-is-the-future-of-net/), but if I remember correctly, we need to build another loader that is built with .NET Core if we decide to port v3 to .NET Core. Even If you allow to do, I'll wait for the release of .NET Core 3.0 though, because (its release date is scheduled for September 23, 2019)[https://dotnet.microsoft.com/platform/support/policy/dotnet-core].

**Comment by crosire** [comment]
I don't think .NET Core is an option for SHVDN unfortunately. Too much potential to break existing scripts, no C++/CLI, ...
And I really don't want two separate projects (v3 in Core and v2 in Framework), that's why v3 got stuck in the first place.

**Comment by kagikn** [comment]
I respect your opinion. If there were a way to run .NET Framework assemblies and .NET Core assemblies without breaking compatibility in one loader, you would be immediately convinced, but that's not the case. So I won't try to port v3 to .NET Core unless there are a solution that convinces you or you change your opinion.

**Comment by crosire** [comment]
Merged into master now.

I kinda want to have the v2 API in C# too for consistency, so did a port here: https://github.com/crosire/scripthookvdotnet/tree/csharp_port.
There is one issue I ran into however: `GTA.UI.WIDTH` and `GTA.UI.HEIGHT` were defined as `static const int` in C++/CLI, which translates to a `static modopt(System.Runtime.Compilerservices.IsConst) int` in IL. There doesn't seem to be an equivalent in C#, I can only write `static int` or `const int`, both of which generate different type signatures. Therefore any scripts that make use of these two constants do not work with the C# port and throw a `MissingFieldException`. If anybody has any idea on how to solve it (e.g. somehow emit the `modopt` via an attribute or something), I'd appreciate it!
